proto/usr/share/initramfs-tools/scripts/triton

```diff
@@ -0,0 +1,70 @@
+#!/bin/sh
+
+# This Source Code Form is subject to the terms of the Mozilla Public
+# License, v. 2.0. If a copy of the MPL was not distributed with this
+# file, You can obtain one at http://mozilla.org/MPL/2.0/.
+#
+
+#
+# Copyright 2020 Joyent, Inc.
+#
+
+#
+# This script provides helper functions for setting up Triton.
+#
+
+################################################################################
+# Set up logging.
+# Don't change this unless you absolutely need to. Don't be fancy.
+# It's pretty fragile.
+# Doing the wrong thing here can result in the image being unbootable, or
+# all the scripts failing. Don't rely on anything. At this point in the boot
+# process you're only in the initrd image. Very little of what you expect is
+# available. As noted above, this runs in /bin/sh, which is *dash*. Don't use
+# bashisms. They will not work.
+################################################################################
+set -o xtrace
+
+TRACE_LOG=${rootmnt:=/}/var/log/triton/boot/$(basename "$0").log
+#TRACE_LOG=${rootmnt:=/}/$(basename "$0").log
+mkdir -p "$(dirname "$TRACE_LOG")"
+exec 1>>"${TRACE_LOG}"
+exec 2>>"${TRACE_LOG}"
+PS4=
+PS4="${PS4}"'['"$0"'] LINE:${LINENO}: '
+export PS4
+################################################################################
+# End extremely fragile portion
+################################################################################
+
+# Make sure we don't get tripped up on these being unset.
+: "${quiet:=false}"
+: "${rootmnt:=/}"
+
+# Setup the config dir.
+#
+# Before `sdc-server setup` there is no /opt directory, so we use an in memory
+# config dir under /usr/triton/config.
+# Once `sdc-server setup` completes, the temporary /usr/triton/config contents
+# will be copied to the persisted /opt/triton/config directory, and a symlink
+# is then used to the real persisted directory.
+# All triton tooling/software should use /usr/triton/config/, which should
+# always work regardless of persistence.
+getTritonConfigDir() {
+    zconfig=$(zfs mount | grep '/opt' >/dev/null || echo "not found")
+    triton_config_dir="${rootmnt}/opt/triton/config"
+    usr_triton_config_dir="${rootmnt}/usr/triton/config"
+
```

Comment:
Will fix

Reply:
Ok.

Reply:
Yeah, we create most of the same datasets in the `joysetup.sh` script, including `zones/opt` which will be mounted on `/opt`. A set up LinuxCN will necessarily have `/opt`. Even if everything else exists, if the Linux CN boots without `zones/opt` mounted on `/opt` it will show up as unsetup in CNAPI.

It gets created here: https://github.com/joyent/sdc-headnode/blob/linuxcn/scripts/joysetup.sh#L677-L687

Currently LinuxCN doesn't have a non-triton setup, but I'd like to have one at some point. I'd like to be able to run LinuxCN (Triton Linux? SmartLX?) as a standalone because even for SmartOS being able to spawn one ad hoc makes a lot of debugging easier without having to have a full Triton set up for it.

Reply:
Okay, beautiful! Make sure you mention that joysetup should've created zones/opt for the casual reader of this script.